### DISCUSSION TITLE
Failed to connect to spark

### DISCUSSION THREAD

**Discussion by willim32533** [apache/airflow#20635] [Q&A]
Very new to Airflow and still trying to work out the right way to trigger spark jobs using Airflow.   Using the simple DAG below with default connection spark_default pointing to spark://spark:7070 is unable to connect to spark master (log messages below) even though the IP address that it gets is the IP address of the master spark container.  The airflow and spark containers are all part of the same docker compose, and are all part of the airflow-default bridge network.  (Docker desktop on Windows environment)

Any ideas whether this is an Airflow config, Spark config, or Docker config issue?

```"""
Example Airflow DAG to submit Apache Spark applications using `SparkSubmitOperator`.
"""
from datetime import datetime

from airflow.models import DAG
from airflow.providers.apache.spark.operators.spark_submit import SparkSubmitOperator

with DAG(
    dag_id='example_spark_operator',
    schedule_interval=None,
    start_date=datetime(2021, 1, 1),
    catchup=False,
    tags=['example'],
) as dag:
    # [START howto_operator_spark_submit]
    submit_job = SparkSubmitOperator(
        application="/opt/spark/examples/src/main/python/pi.py", task_id="submit_job"
    )
```

> [2021-12-31, 00:08:15 UTC] {spark_submit.py:514} INFO - 21/12/31 00:08:15 INFO Utils: Successfully started service 'sparkDriver' on port 35347.

> [2021-12-31, 00:08:15 UTC] {spark_submit.py:514} INFO - 21/12/31 00:08:15 INFO Utils: Successfully started service 'SparkUI' on port 4040.

> [2021-12-31, 00:08:15 UTC] {spark_submit.py:514} INFO - 21/12/31 00:08:15 INFO SparkUI: Bound SparkUI to 0.0.0.0, and started at http://421f90bc9431:4040

> [2021-12-31, 00:08:15 UTC] {spark_submit.py:514} INFO - 21/12/31 00:08:15 INFO StandaloneAppClient$ClientEndpoint: Connecting to master spark://spark:7070...

> [2021-12-31, 00:08:15 UTC] {spark_submit.py:514} INFO - 21/12/31 00:08:15 WARN StandaloneAppClient$ClientEndpoint: Failed to connect to master spark:7070

> [2021-12-31, 00:08:15 UTC] {spark_submit.py:514} INFO - org.apache.spark.SparkException: Exception thrown in awaitResult:

> [2021-12-31, 00:08:15 UTC] {spark_submit.py:514} INFO - at org.apache.spark.deploy.client.StandaloneAppClient$ClientEndpoint$$anon$1.run(StandaloneAppClient.scala:107)

> [2021-12-31, 00:08:15 UTC] {spark_submit.py:514} INFO - Caused by: java.io.IOException: Failed to connect to spark/172.22.0.9:7070

**Comment by potiuk** [comment]
Likely spark is only configured to listen to port 7070 (or is bound to localhost and not 0.0.0.0).

**Reply by willim32533** [reply]
@potiuk Thank you for the reply.  Any idea as to how to address that?

**Reply by potiuk** [reply]
Make it listen ?  No idea. I have no spark experience.

**Comment by demisomi** [comment]
please @willim32533  were you able to find a solution to the problem

**Comment by YashPayU** [comment]
I'm running Airflow locally and Standalone Spark cluster in my local. While submitting the job from airflow, I'm getting connection issue. Any help would be really appreciated.

Apache Airflow: apache/airflow:2.6.3
Apache Spark: spark-3.4.1
JAVA JDK: jdk-20

**Reply by Madhukrishna-CD** [reply]
Is both spark and airflow installed in the same servers or different

**Reply by YashPayU** [reply]
Both are deployed locally using containers. I'm able to submit spark jobs from spark worker to spark master but unable to submit the jobs either from macbook local terminal or from airflow deployed locally in a container.

**Reply by YashPayU** [reply]
Instead of creating spark standalone cluster from docker image, I tried creating standalone cluster from shell scripts available in spark zip file and now I'm able to submit spark jobs from my local. But i'm still unable to connect standalone master with airflow deployed locally.

I've extended my airflow image with below dockerfile and used it as a base image for officially available docker-compose file

FROM apache/airflow:2.2.5-python3.8
USER root

# Install OpenJDK-11
RUN apt update &&
apt-get install -y openjdk-11-jdk &&
apt-get install -y ant &&
apt-get clean;

# Set JAVA_HOME
ENV JAVA_HOME /usr/lib/jvm/java-11-openjdk-amd64/
RUN export JAVA_HOME

USER airflow

COPY ./requirements.txt /
RUN pip install -r /requirements.txt

COPY --chown=airflow:root ./dags /opt/airflow/dags
